<a href="https://colab.research.google.com/github/caalr11/FisicaComputacional1/blob/master/Sprint_3_Ejercicios_Parte_1_SV.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Examples:

#### Importar pandas:

In [ ]:
import pandas as pd
## import pandas -> Aquí importamos la librería
## as pd -> Aquí renombramos la librería como "pd". Esto nos ahorra tiempo en el uso, dado que es un alias corto.

In [ ]:
pd.read_csv()

### Leamos una tabla en formato csv:

**En Google Colab**

In [ ]:
#### Ejecutamos y subimos el archivo
from google.colab import files
uploaded = files.upload()

Saving Telco-Customer-Churn-Dirty.csv to Telco-Customer-Churn-Dirty.csv


In [ ]:
# Leemos el archivo:
#df = pd.read_csv("Telco-Customer-Churn-Dirty.csv",delimiter=',')
df = pd.read_csv("Telco-Customer-Churn-Dirty.csv",sep = ',')

In [ ]:
df

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female            0.0     Yes         No     1.0   
1     5575-GNVDE    Male            0.0      No        NaN    34.0   
2            NaN    Male            0.0      No         No     2.0   
3     7795-CFOCW    Male            0.0      No         No    45.0   
4     9237-HQITU  Female            0.0      No         No     2.0   
...          ...     ...            ...     ...        ...     ...   
7390  2294-DMMUS  Female            NaN     Yes        Yes    62.0   
7391  5921-NGYRH     NaN            NaN     Yes         No    64.0   
7392  1897-OKVMW  Female            NaN     Yes        Yes    64.0   
7393  5117-IFGPS    Male            NaN     Yes         No    29.0   
7394  3866-MDTUB  Female            NaN      No         No     2.0   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0              No  No phone service             DSL             No  ...   
1             Yes                No             DSL            Yes  ...   
2             Yes                No             DSL            Yes  ...   
3              No  No phone service             DSL            Yes  ...   
4             Yes                No     Fiber optic             No  ...   
...           ...               ...             ...            ...  ...   
7390          Yes               Yes             DSL             No  ...   
7391          Yes                No     Fiber optic            Yes  ...   
7392          Yes               Yes             DSL            NaN  ...   
7393          Yes               Yes             NaN             No  ...   
7394          Yes                No             DSL            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No             NaN   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7390               No          No         Yes             Yes  Month-to-month   
7391               No          No          No              No  Month-to-month   
7392              Yes         Yes         Yes             Yes        Two year   
7393               No          No          No             Yes  Month-to-month   
7394               No          No         Yes             Yes  Month-to-month   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0                 Yes                        NaN          29.85         29.85   
1                  No               Mailed check          56.95        1889.5   
2                 Yes               Mailed check          53.85        108.15   
3                  No  Bank transfer (automatic)          42.30       1840.75   
4                 Yes           Electronic check          70.70        151.65   
...               ...                        ...            ...           ...   
7390              Yes    Credit card (automatic)          70.45       4300.45   
7391              Yes               Mailed check          73.55          4764   
7392              Yes               Mailed check          90.60       5817.45   
7393              Yes           Electronic check          84.30       2357.75   
7394              Yes           Electronic check          70.75         146.9   

     Churn  
0       No  
1       No  
2      Yes  
3       No  
4      Yes  
...    ...  
7390    No  
7391    No  
7392    No  
7393    No  
7394   Yes  

[7395 rows x 21 columns]

**En vscode o local:**

In [ ]:
# Leemos el archivo:
df = pd.read_csv(r"url_de_mi_computador_donde_vive_mi_archivo/music_project_en.csv")

# df = pd.read_csv(r"C:\Users\OMEN\Downloads\music_project_en.csv")

In [ ]:
df.size

df.shape[1]

7395

In [ ]:
(df.isna().sum())

customerID           371
gender               372
SeniorCitizen       1056
Partner              376
Dependents           372
tenure               742
PhoneService         367
MultipleLines        365
InternetService      372
OnlineSecurity       375
OnlineBackup         367
DeviceProtection     370
TechSupport          381
StreamingTV          364
StreamingMovies      370
Contract             375
PaperlessBilling     359
PaymentMethod        370
MonthlyCharges       741
TotalCharges         371
Churn                378
dtype: int64

## Visualicemos los datos:


1.   `df.head(10)`: Imprime solo las 10 primeras filas del Dataframe.
2.   `df.columns`: Nombre de las columnas.
3.   `df['col_1']`: Selecciono solo la columna llamada "col_1".
4.   `df.info()`: Tipo de datos de las columnas(int,float,object) junto con otros datos interesantes.
5.   `df.isna().sum()`: Cantidad de valores faltantes por columna.
6.   `df['col_1'].fillna(0)`: Relleno los datos faltantes de mi columna 'col_1' con el valor 0.
7.   `df.duplicated().sum()`: Cantidad de valores duplicados.
8.   `df.drop_duplicated()`: Elimina filas duplicadas.
9.   `df.loc[]`: Selecciona columnas y filas con sus nombres.
10.   `df.iloc[]`: Selecciona columnas y filas con índices.
11.   `df['col_1'].nunique()`: cuantos valores unicos poseo en mi columna "col_1".

**Adicionalmente:**

1.   `df[df['genero']=='Otro']`: Filtramos nuestro dataframe solo las filas que contengan la columna llamada "genero" el valor "Otro'.
2.   `df[ (df['genero']=='Otro') & (df['edad']>50) ]`: Filtramos nuestro dataframe solo las filas que contengan la columna llamada "genero" el valor "Otro' y además en la columna "edad" contengan el valor 50

# Exercises for Live session 1

Ejercicio 1: Carga el archivo CSV con los datos alterados y muestra las primeras 10 filas del DataFrame.

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv("Telco-Customer-Churn-Dirty.csv",sep = ',')

In [ ]:
df.head(10)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female            0.0     Yes         No     1.0           No   
1  5575-GNVDE    Male            0.0      No        NaN    34.0          Yes   
2         NaN    Male            0.0      No         No     2.0          Yes   
3  7795-CFOCW    Male            0.0      No         No    45.0           No   
4  9237-HQITU  Female            0.0      No         No     2.0          Yes   
5  9305-CDSKC  Female            0.0      No         No     8.0          Yes   
6         NaN    Male            0.0      No        Yes    22.0          Yes   
7  6713-OKOMC  Female            0.0      No         No    10.0           No   
8  7892-POOKP  Female            0.0     Yes         No     NaN          Yes   
9  6388-TABGU    Male            0.0      No        Yes    62.0          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   
5               Yes     Fiber optic             No  ...              Yes   
6               Yes     Fiber optic             No  ...               No   
7  No phone service             DSL            Yes  ...               No   
8               Yes     Fiber optic             No  ...              Yes   
9                No             DSL            Yes  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No             NaN              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   
5          No         Yes             Yes  Month-to-month              Yes   
6          No         Yes              No  Month-to-month              Yes   
7          No          No              No  Month-to-month               No   
8         Yes         Yes             Yes  Month-to-month              Yes   
9          No          No              No        One year               No   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0                        NaN          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  
5           Electronic check          99.65         820.5   NaN  
6    Credit card (automatic)          89.10        1949.4    No  
7               Mailed check          29.75         301.9    No  
8           Electronic check         104.80           NaN   Yes  
9  Bank transfer (automatic)          56.15       3487.95    No  

[10 rows x 21 columns]

Ejercicio 2. Verifica el tipo de datos de cada columna del dataset.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7395 entries, 0 to 7394
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7024 non-null   object 
 1   gender            7023 non-null   object 
 2   SeniorCitizen     6339 non-null   float64
 3   Partner           7019 non-null   object 
 4   Dependents        7023 non-null   object 
 5   tenure            6653 non-null   float64
 6   PhoneService      7028 non-null   object 
 7   MultipleLines     7030 non-null   object 
 8   InternetService   7023 non-null   object 
 9   OnlineSecurity    7020 non-null   object 
 10  OnlineBackup      7028 non-null   object 
 11  DeviceProtection  7025 non-null   object 
 12  TechSupport       7014 non-null   object 
 13  StreamingTV       7031 non-null   object 
 14  StreamingMovies   7025 non-null   object 
 15  Contract          7020 non-null   object 
 16  PaperlessBilling  7036 non-null   object 


Ejercicio 3: Verifica si hay valores ausentes en el conjunto de datos y reemplázalos adecuadamente.

In [ ]:
df.isna().sum()

customerID           371
gender               372
SeniorCitizen       1056
Partner              376
Dependents           372
tenure               742
PhoneService         367
MultipleLines        365
InternetService      372
OnlineSecurity       375
OnlineBackup         367
DeviceProtection     370
TechSupport          381
StreamingTV          364
StreamingMovies      370
Contract             375
PaperlessBilling     359
PaymentMethod        370
MonthlyCharges       741
TotalCharges         371
Churn                378
dtype: int64

In [ ]:
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7395 entries, 0 to 7394
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7024 non-null   object 
 1   gender            7023 non-null   object 
 2   SeniorCitizen     6339 non-null   float64
 3   Partner           7019 non-null   object 
 4   Dependents        7023 non-null   object 
 5   tenure            6653 non-null   float64
 6   PhoneService      7028 non-null   object 
 7   MultipleLines     7030 non-null   object 
 8   InternetService   7023 non-null   object 
 9   OnlineSecurity    7020 non-null   object 
 10  OnlineBackup      7028 non-null   object 
 11  DeviceProtection  7025 non-null   object 
 12  TechSupport       7014 non-null   object 
 13  StreamingTV       7031 non-null   object 
 14  StreamingMovies   7025 non-null   object 
 15  Contract          7020 non-null   object 
 16  PaperlessBilling  7036 non-null   object 


In [ ]:
df = df.fillna(0)

In [ ]:
df.isna().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

Ejercicio 4: Determina si hay filas duplicadas en el conjunto de datos y elimínalas si existen.

In [ ]:
df.duplicated().sum()

np.int64(352)

In [ ]:
#df = df.drop_duplicates()
df.drop_duplicates(inplace=True)

In [ ]:
df.duplicated().sum()

np.int64(0)

Ejercicio 5: Renombra las columnas del DataFrame para que sigan un formato en minúsculas y sin espacios.

In [ ]:
df.columns = df.columns.str.lower().str.strip().str.replace('techsupport','techsupport1')

In [ ]:
df.columns

Index(['customerid', 'gender', 'seniorcitizen', 'partner', 'dependents',
       'tenure', 'phoneservice', 'multiplelines', 'internetservice',
       'onlinesecurity', 'onlinebackup', 'deviceprotection', 'techsupport1',
       'streamingtv', 'streamingmovies', 'contract', 'paperlessbilling',
       'paymentmethod', 'monthlycharges', 'totalcharges', 'churn'],
      dtype='object')

Ejercicio 7: Usa loc para seleccionar información de un cliente específico ('7590-VHVEG') y iloc para obtener los primeros 5 registros del DataFrame.

Ejercicio 8: Cuenta cuántos valores únicos hay en la columna payment_method.

Ejercicio 9: Filtra los clientes que tienen un contrato de "Month-to-month" y pagan más de 80 en monthly_charges.

In [ ]:
df_filtrado = df[(df['contract'] == "Month-to-month") & (df['monthlycharges'] > 80 )]

In [ ]:
df_filtrado.describe()

seniorcitizen       tenure  monthlycharges  totalcharges
count    1223.000000  1223.000000     1223.000000   1223.000000
mean        0.268193    23.319706       92.900818   2350.334342
std         0.443200    20.081951        8.184004   1955.183808
min         0.000000     0.000000       80.050000      0.000000
25%         0.000000     5.000000       85.600000    652.575000
50%         0.000000    19.000000       92.550000   1885.150000
75%         1.000000    38.000000       99.450000   3756.425000
max         1.000000    72.000000      117.450000   8061.500000